### PR TITLE
Feature/8232 tag settings

### LINKED ISSUE #8232: Feature: Tag Settings
Sharing here some design inspiration by @mattmiklic.  Original at p9c7Jx-do-p2.

> I’ve been working on a native Tags settings design for iOS and Android. To refresh, this is what the UI looks like on mobile web in Calypso:
> tag-settings-calypso
>
> Here’s how I’ve translated that to native 
>
> tag-settings-native
>
> Notes:
> - The basic flow remains the same.
> - I’ve eliminated the folder icons as they seemed to only be decorative, without offering any real benefit.
> - The circled number that denotes the number of posts is tappable. On iOS this would work similarly to a detail disclosure button. On Android it would be a supplemental action. Tapping the number opens the list of posts using the existing post list. Tapping anywhere outside the circled number opens the edit view. This behavior is in line with the OS specs for disclosure buttons & supplemental actions.
> - Rather than show a 0 on every tag that doesn’t have any posts associated with it, I’m proposing that we just not show a number for those tags. The repeating 0 seems like it doesn’t really add much value.
> - Tapping the trash icon on the detail view opens an alert confirming the deletion; the message and button options should be the same as Calypso. If feasible, on iOS, we can add a swipe-left gesture on the tags list to reveal a “Delete” option.
> - The search form for iOS shown here is different from what we’ve used in the past. But this matches the new pattern in iOS 11 for a combined toolbar + search field. When this is being implemented I can flesh out the details on color, etc. as needed.
> - I didn’t include a detailed view of search results; I can if needed, but I assume they’ll be mostly self-explanatory

### PR DESCRIPTION
**Pull request by aerych**
Closes #8232 

This is a pr to merge the Tags Settings feature written by @ctarda.   The work was reviewed in several installments and merged into this feature branch. See: #8236 #8244 #8248 #8278 #8288 #8305 
I would like a final test and review of the merged branch (for paranoia) of the feature branch before merging to develop. 

To test:
Be logged into wpcom. 
- Choose a site from My Sites and view its Settings > Tag Settings screen
- Confirm you can add new tags
- Confirm you can edit existing tags
- Confirm you can delete tags by swiping in the list or from the button in the detail view. 
- Confirm the list is updated in response to a newly created tag or a deleted tag.
- Confirm that tag descriptions are being updated.

Bonus, test with a self-hosted site (sans-Jetpack).  You'll need to edit the SiteSettingsViewController to view the Writing section in order to do this.  We've opted to leave the feature hidden from self-hosted for now but the support is there.

@elibud - fyi as this is relevant to your interests.

### CODE REVIEW

**Comment by elibud** [comment]
Wow, I'm so happy to see this implemented 🎉 great work @ctarda!

I did some initial testing and noticed that unless we navigate to the tags screen the value displayed on the site settings screen is 0. I think this is someone we want to tackle before shipping this one, so we can review and merge next week, making a note to whether:

a. Automatically fetch the count of tags on the Site Settings screen, or
b. Don't display the number of tags unless it's up to date, or
c. Don't display the number of tags at all

Before ever fetching tags | After fetching tags
--- | ---
simulator screen shot - iphone 8 - 2017-12-15 at 11 01 52 |simulator screen shot - iphone 8 - 2017-12-15 at 11 16 46

I think that to avoid unnecessary network calls to fetch information that might not be relevant at all users the more reasonable option is c. @aerych @ctarda what do you think?

**Comment by aerych** [comment]
> I think that to avoid unnecessary network calls to fetch information that might not be relevant at all users the more reasonable option is c.

I was going to suggest we hide the count if its zero and show it otherwise but that still leaves the possibility for it to reflect an inaccurate count.  @mattmiklic is there a strong need to show the number of tags on the settings screen or would you be cool with removing it from here.

**Comment by aerych** [comment]
@elibud I think the merge issues should be resolved now.  I've opted to go ahead and hide the count but we can revisit later pending @mattmiklic's thoughts.

**Comment by mattmiklic** [comment]
> @mattmiklic is there a strong need to show the number of tags on the settings screen or would you be cool with removing it from here.

It seems fine to remove the count there.

**Comment by elibud** [comment]
I noticed some odd behavior when searching:

tagoddities

Should we disable search and ship what we have or is this something easily addressed? @aerych @ctarda what do you think?

File: WordPress/Classes/ViewRelated/Blog/SiteManagement/SiteTagsViewController.swift [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return resultsController.sections?[section].numberOfObjects ?? 0
+    }
+
+    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        guard let cell = tableView.dequeueReusableCell(withIdentifier: TableConstants.cellIdentifier, for: indexPath) as? TitleBadgeDisclosureCell, let tag = tagAtIndexPath(indexPath) else {
+            return TitleBadgeDisclosureCell()
+        }
+
+        cell.name = tag.name
```
Comment (nheagy) [inline WordPress/Classes/ViewRelated/Blog/SiteManagement/SiteTagsViewController.swift]: I think this should have html entities decoded? I did a test with a tag name containing an `&` and it became `&amp;` and displayed that way in the UI, but not on the web.

fwiw, this seems to also be an existing problem with the tags list in the blog posts list.

File: WordPress/Classes/ViewRelated/Blog/SiteManagement/SiteTagsViewController.swift [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+        }
+
+        return cell
+    }
+
+    fileprivate func tagAtIndexPath(_ indexPath: IndexPath) -> PostTag? {
+        return resultsController.object(at: indexPath) as? PostTag
+    }
+
+    func controllerDidChangeContent(_ controller: NSFetchedResultsController<NSFetchRequestResult>) {
```
Comment (nheagy) [inline WordPress/Classes/ViewRelated/Blog/SiteManagement/SiteTagsViewController.swift]: This is a method from `NSFetchedResultsControllerDelegate`, but it's in the section marked `Table view datasource`. Thoughts on reorganizing this a bit to clean up the layout of these methods?

File: WordPress/Classes/ViewRelated/Blog/SiteManagement/SiteTagsViewController.swift [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+            return
+        }
+
+        delete(selectedTag)
+    }
+
+    private func delete(_ tag: PostTag) {
+        let tagsService = PostTagService(managedObjectContext: ContextManager.sharedInstance().mainContext)
+        refreshControl?.beginRefreshing()
+        tagsService.delete(tag, for: blog, success: {
```
Comment (nheagy) [inline WordPress/Classes/ViewRelated/Blog/SiteManagement/SiteTagsViewController.swift]: should this closure use `[weak self]`? I couldn't see a problem while running the app, but seems like it _could_ be a retain loop?

File: WordPress/Classes/ViewRelated/Blog/SiteSettingsViewController.m [hunk 4]
```diff
@@ -974,6 +1000,18 @@ - (void)loginValidationFailedWithError:(NSError *)error
     }
 }
 
+- (NSString *)getTagsCountPresentableString:(NSInteger)tagCount
+{
+    NSString *format = NSLocalizedString(@"%@ Tags", @"The number of tags in the writting settings. Plural. %@ is a placeholder for the number");
+    
+    if (tagCount == 1) {
```
Comment (nheagy) [inline WordPress/Classes/ViewRelated/Blog/SiteSettingsViewController.m]: Just making a note here that this is the English-centric way of pluralizing, however, I don't think we're doing any better anywhere else in the code 😜

**Comment by aerych** [comment]
Thanks @nheagy!  Comments addressed. :)
@elibud would you take a look at that last commit, and then retest for the search behavior you noticed.  It should be resolved but the fix I came up with is rather hackish.  I tried a few other things but nothing looked as clean from the users pov.

**Comment by elibud** [comment]
@aerych I've added an extra param `isDestructiveAction` to `SettingsTitleSubtitleController` to show the action as destructive:

simulator screen shot - iphone 8 - 2017-12-19 at 09 41 43

Let me know what you think about that, I can revert it if it's not popular ;)

I'm still having issues with search though, when searching for a specific tag, and then deleting it I get the No Results View, which shouldn't be displayed in that situation:

tags

**Comment by nheagy** [comment]
The `isDestructiveAction` changes look good to me, fwiw

**Comment by elibud** [comment]
@koke it doesn't look like @aerych will have time to work on this one before he abandons us, so we are going to have to take it :)

**Comment by aerych** [comment]
@elibud, surely you jest ;)
See if 93e3a48 does the trick.

File: MIGRATIONS.md [hunk 5]
```diff
@@ -4,6 +4,8 @@ This file documents changes in the data model. Please explain any changes to the
 data model as well as any custom migrations.
 
 ## WordPress 68
+- @ctarda 2017-11-27
+- `PostTag` added `tagDescription`  string property and `postCount` integer property. Store an optional description and the number of posts a tag has been used in.
```
Comment (koke) [inline MIGRATIONS.md]: If this is for 9.2, I think we need a new data model version

**Comment by elibud** [comment]
@aerych yes it's fixed 🏅 great job! As @koke mentioned we need to update this from develop and add a new data model version since `68` is taken now, let me know if you want me to tackle that.

**Comment by aerych** [comment]
Revision created. There are a few commits in the history that have revision 68 with the new fields for PostTags but I'm reluctant to squash them.

**Comment by elibud** [comment]
That's 👍 with me :shipit:

**Comment by aerych** [comment]
Thanks all! :)

**Pull request merged**